config/proxy/proxy.go

```diff
@@ -0,0 +1,197 @@
+	obfuscate bool
+
+	// listeners is a list of functions to call when the config changes.
+	listeners []func(old, new *ProxyConfig)
+	mu        sync.Mutex
+}
+
+func Init(saveDir string, obfuscate bool) (*config, error) {
+	if saveDir == "" {
+		saveDir = defaultConfigSaveDir
+		_config.filePath = filepath.Join(saveDir, defaultConfigFilename)
+	}
+
+	_config.obfuscate = obfuscate
+
+	saved, err := readExistingConfig(_config.filePath, obfuscate)
+	if err != nil {
+		log.Errorf("Failed to read existing client config: %w", err)
+	}
+
+	_config.config.Set(saved)
+	notifyListeners(nil, saved)
+	return _config, nil
+}
+
+// GetConfig implements services.ConfigHandler
+func (c *config) GetConfig() *ProxyConfig {
+	conf, _ := c.config.Get(eventual.DontWait)
+	return conf.(*ProxyConfig)
+}
+
+// SetConfig implements services.ConfigHandler
+func (c *config) SetConfig(new *ProxyConfig) {
+	old := c.GetConfig()
+	c.config.Set(new)
+	if err := saveConfig(c.filePath, new, c.obfuscate); err != nil {
+		log.Errorf("Failed to save client config: %w", err)
+	}
+
+	notifyListeners(old, new)
+}
+
+// GetConfig returns the current client config. An error is returned if the config is not available
+// within the given timeout.
+func GetConfig(timeout time.Duration) (*ProxyConfig, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	conf, err := _config.config.Get(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get config: %w", err)
+	}
+
+	return conf.(*ProxyConfig), nil
+}
+
+// GetProToken returns the pro token from the current client config. An error is returned if the
+// config is not available within the given timeout.
+func GetProToken(timeout time.Duration) (string, error) {
+	config, err := GetConfig(timeout)
+	if err != nil || config == nil {
+		return "", err
+	}
+
+	return config.GetProToken(), nil
+}
+
+// GetProxyConfigs returns the list of proxy configs from the current client config. An error is
+// returned if the config is not available within the given timeout.
+func GetProxyConfigs(timeout time.Duration) ([]*apipb.ProxyConnectConfig, error) {
+	config, err := GetConfig(timeout)
+	if err != nil || config == nil {
+		return nil, err
+	}
+
+	return config.GetProxy().GetProxies(), nil
+}
+
+// readExistingConfig reads a config from a file at the specified path, filePath,
+// deobfuscating it if obfuscate is true.
+func readExistingConfig(filePath string, obfuscate bool) (*ProxyConfig, error) {
+	infile, err := os.Open(filePath)
+	if err != nil {
+		return nil, fmt.Errorf("unable to open config file %v for reading: %w", filePath, err)
+	}
+	defer infile.Close()
+
+	var in io.Reader = infile
+	if obfuscate {
+		in = rot13.NewReader(infile)
+	}
+
+	bytes, err := io.ReadAll(in)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read config from %v: %w", filePath, err)
+	}
+
+	if len(bytes) == 0 {
+		return nil, nil // file is empty
```

Comment:
This is how it was in the original code so I left it. I think it treats an empty file as if the file didn't exist. It might be better to return an empty `ProxyConfig` instead.